### PR TITLE
feat(api): parameters

### LINKED ISSUE #1560: Specify explicit parameters from inside python code
Users should be able to declare parameters inside a script as such, without passing them on the command line. This would later allow us to show them on the KG and open other use-cases like showing changes to parameters between runs.

This might take the form of 
```
lr = renku.parameter('learning_rate', 0.9)
```

### PR DESCRIPTION
**Pull request by mohammad-sdsc**
# Description

Documentation for API.
Allows defining run parameters for workflows.

Fixes #1560

### CODE REVIEW

File: renku/core/models/workflow/converters/cwl.py [hunk 1]
```diff
@@ -279,6 +279,34 @@ def _convert_step(step, tmpdir, basedir, filename=None):
...
+                "input_renku_metadata",
+                param_type="Directory",
+                input_binding=None,
+                default={"path": os.path.abspath(os.path.join(basedir, ".renku")), "class": "Directory"},
+            )
+        )
+
+        workdir_req.listing.append(
+            cwlgen.InitialWorkDirRequirement.Dirent(
+                entry="$(inputs.input_git_directory)", entryname=".git", writable=False
```
Comment (Panaetius) [inline renku/core/models/workflow/converters/cwl.py:298]: Did you check what the performance implication is of copying over the `.git` directory for each workflow rerun? Especially on large projects.

File: renku/core/models/workflow/parameters.py [hunk 2]
```diff
@@ -265,6 +265,53 @@ def as_jsonld(self):
...
+
+    value = attr.ib(default=None, type=str, kw_only=True)
+
+    type = attr.ib(default=None, type=str, kw_only=True)
+
+    @staticmethod
+    def generate_id(run_id, name):
+        """Generate an id."""
+        name = urllib.parse.quote(name, safe="")
+        return "{}/arguments/{}".format(run_id, name)
```
Comment (Panaetius) [inline renku/core/models/workflow/parameters.py:286]: should we change the id to `[...]/parameter/[...]` ?

**Review by Panaetius — CHANGES_REQUESTED** [review]
This looks really nice, thank you! I have a minor change request, otherwise just some questions.

Right now a user could not change the value of a Parameter on `renku update`/`rerun`, correct? Only by modifying the script, but not e.g. passing it on the cmdline `renku update --param param-2 41`? Should we add something like this? It might be error prone if two scripts define `param-2` but then again, maybe a user wants to set the same value in multiple places.

File: renku/core/models/workflow/converters/cwl.py (hunk 1 above)
Comment (mohammad-sdsc) [reply renku/core/models/workflow/converters/cwl.py:298]: It creates only a symlink and does not do an actual copy. This is not needed once we can load metadata without traversing git history. I'll add a TODO to remove when https://github.com/SwissDataScienceCenter/renku-python/issues/1043 is done.

File: renku/core/models/workflow/parameters.py (hunk 2 above)
Comment (mohammad-sdsc) [reply renku/core/models/workflow/parameters.py:286]: Was waiting for a second thought on that 😄

**Comment by mohammad-sdsc** [comment]
> Right now a user could not change the value of a Parameter on `renku update`/`rerun`, correct? Only by modifying the script, but not e.g. passing it on the cmdline `renku update --param param-2 41`? Should we add something like this? It might be error prone if two scripts define `param-2` but then again, maybe a user wants to set the same value in multiple places.

I believe this requires more thought and I prefer to keep the use-case limited to API for this story. I'll create a follow up story to enable users to pass parameters via command line.

**Review by Panaetius — APPROVED** [review]
Thank you!

**Pull request merged**